sql/core/src/main/scala/org/apache/spark/sql/columnar/CachedBatchSerializer.scala

```diff
@@ -0,0 +1,291 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.columnar
+
+import org.apache.spark.annotation.{DeveloperApi, Since}
+import org.apache.spark.internal.Logging
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.dsl.expressions._
+import org.apache.spark.sql.catalyst.expressions.{And, Attribute, AttributeReference, BindReferences, EqualNullSafe, EqualTo, Expression, GreaterThan, GreaterThanOrEqual, In, IsNotNull, IsNull, Length, LessThan, LessThanOrEqual, Literal, Or, Predicate, StartsWith}
+import org.apache.spark.sql.execution.SparkPlan
+import org.apache.spark.sql.execution.columnar.{ColumnStatisticsSchema, PartitionStatistics}
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.types.{AtomicType, BinaryType, StructType}
+import org.apache.spark.sql.vectorized.ColumnarBatch
+
+/**
+ * Basic interface that all cached batches of data must support. This is primarily to allow
+ * for metrics to be handled outside of the encoding and decoding steps in a standard way.
+ */
+@DeveloperApi
+@Since("3.1.0")
+trait CachedBatch {
+  def numRows: Int
+  def sizeInBytes: Long
+}
+
+/**
+ * Provides APIs for compressing, filtering, and decompressing SQL data that will be
+ * persisted/cached.
+ */
+@DeveloperApi
+@Since("3.1.0")
+trait CachedBatchSerializer extends Serializable {
```

Comment:
I'd like to understand the API design concerns that lead to this design. IMO the two major concerns should be
1. the input to the cache builder should be row-based or columnar?
2. the output of the cache builder should be row-based or columnar?

It seems to me that `convertForCache` is to address 1. `SparkPlan` provides APIs to get row-based or columnar data. But I think exposing `SparkPlan` is overkill: it's a very big class that has many methods. Can we create a trait to abstract the main functionalities?

For 2, the current API looks reasonable.

Reply:
The major goal was to expose some way to provide custom code for compressing/decompressing the cached data.  Decompression was already really fast and had APIs for columnar output vs non-columnar output so I mostly just kept them in place. 

`convertForCache` does have a goal to be able to let someone process columnar input, but because of how the `SparkPlan` is created that is not a simple task. `QueryExecution.executedPlan` which is the source of the `SparkPlan` will always produce row based output. That was one of the design goals for the columnar execution work. It had to be truly transparent. Currently the only way to get around that is by walking the SparkPlan to check if the last stage is a code generation phase with only a child that is doing a columnar to row transition. If so then it is safe to remove that and process it directly as columnar. If you want me to create a trait for `SparkPlan` that exposes a simpler API just for caching, I can, but it will not be able to expose anything columnar unless we make larger changes. I am fine with making those larger changes, I just want to be sure that the community is willing to accept them conceptually before I start to touch a lot of the code.

Reply:
> Currently the only way to get around that is by walking the SparkPlan to check if the last stage is a code generation phase with only a child that is doing a columnar to row transition.

I'd like to understand more about the physical path. So we only pass columnar data to the cache builder if the spark plan can produce columnar data natively (the last operator is a columnar to row transition)?

If that's the case, I'd suggest we add 3 methods: `convertForCache(inpur: RDD[InternalRow])`, `convertForCacheColumnar(inpur: RDD[ColumnarBatch])` and `supportsColumnarInput(schema): Boolean`.

Reply:
Columnar to row transitions are inserted as a part of the `ApplyColumnarRulesAndInsertTransitions` rule in `QueryExecution.preparations`.  It looks for `SparkPlan` instances that output columnar data followed by ones that take row based data as input.  The rule inserts a `ColumnarToRowExec` transition between them.  This transition supports code generation so in normal operation it will then be wrapped in a `WholeStageCodeGenExec`. `ApplyColumnarRulesAndInsertTransitions` assumes that the output of any plan must be row based to conform with all of the other code that already makes that assumption.  It should be a minor change to be able to make that optional so `QueryExecution` could provide a new API that would return a `SparkPlan` that might be columnar and might not.

I can put together an prototype of how that would look if you are OK with it.

Reply:
`SparkPlan` is so internal and I'd like to try harder to avoid leaking it to the developer APIs.

After all, the cache builder only needs an RDD (either contains rows or columnar batches). We should keep the plan execution inside Spark and only exchange data (RDD) with the plugins.